src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_kds.c

```diff
@@ -349,6 +385,226 @@ static int xocl_context_ioctl(struct xocl_dev *xdev, void *data,
+
+	/* Unlock the bitstream for this HW context if no reference is there */
+	(void)xocl_icap_unlock_bitstream(xdev, hw_ctx->xclbin_id);
+
+	ret = kds_free_hw_ctx(client, hw_ctx);
+
+	mutex_unlock(&client->lock);
+
+        return ret;
+}
+
+static int
+xocl_cu_ctx_to_info(struct xocl_dev *xdev, struct drm_xocl_open_cu_ctx *cu_args,
+	struct kds_client_hw_ctx *hw_ctx, struct kds_client_cu_info *cu_info)
+{
+        uint32_t slot_hndl = hw_ctx->slot_idx;
+        struct kds_sched *kds = &XDEV(xdev)->kds;
+        char *kname_p = cu_args->cu_name;
+        struct xrt_cu *xcu = NULL;
+        char iname[64];
+        char kname[64];
+        int i = 0;
+
+        strcpy(kname, strsep(&kname_p, ":"));
+        strcpy(iname, strsep(&kname_p, ":"));
+
+        /* Retrieve the CU index from the given slot */
+        for (i = 0; i < MAX_CUS; i++) {
+                xcu = kds->cu_mgmt.xcus[i];
+                if (!xcu)
+                        continue;
+
+                if ((xcu->info.slot_idx == slot_hndl) &&
+                                (!strcmp(xcu->info.kname, kname)) &&
+                                (!strcmp(xcu->info.iname, iname))) {
+                        cu_info->cu_domain = DOMAIN_PL;
+                        cu_info->cu_idx = i;
+                        goto done;
+                }
+        }
+
+        /* Retrieve the SCU index from the given slot */
+        for (i = 0; i < MAX_CUS; i++) {
+                xcu = kds->scu_mgmt.xcus[i];
+                if (!xcu)
+                        continue;
+
+                if ((xcu->info.slot_idx == slot_hndl) &&
+                                (!strcmp(xcu->info.kname, kname)) &&
+                                (!strcmp(xcu->info.iname, iname))) {
+                        cu_info->cu_domain = DOMAIN_PS;
+                        cu_info->cu_idx = i;
+                        goto done;
+                }
+        }
+
+        return -EINVAL;
+
+done:
+	cu_info->ctx = (void *)hw_ctx;
+        if (cu_args->flags == XOCL_CTX_EXCLUSIVE)
+                cu_info->flags = CU_CTX_EXCLUSIVE;
+        else
+                cu_info->flags = CU_CTX_SHARED;
+
+        return 0;
+}
+
+static inline void
+xocl_close_cu_ctx_to_info(struct drm_xocl_close_cu_ctx *args,
+		   struct kds_client_cu_info *cu_info)
+{
+	/* Extract the CU information */
+	cu_info->cu_domain = get_domain(args->cu_index);
+        cu_info->cu_idx = get_domain_idx(args->cu_index);
+}
+
+int xocl_open_cu_context(struct xocl_dev *xdev, struct drm_file *filp,
+		struct drm_xocl_open_cu_ctx *drm_cu_args)
+{
+	struct kds_client *client = filp->driver_priv;
+	struct kds_client_hw_ctx *hw_ctx = NULL;
+	struct kds_client_cu_ctx *cu_ctx = NULL;
+	struct kds_client_cu_info cu_info = { };
+	int ret = 0;
+
+        mutex_lock(&client->lock);
+
+        hw_ctx = kds_get_hw_ctx_by_id(client, drm_cu_args->hw_context);
+        if (!hw_ctx) {
+                userpf_err(xdev, "No valid HW context is open");
+		mutex_unlock(&client->lock);
```

Comment:
See comments earlier regarding the best practice of having only 1 mutex_lock and 1 mutex_unlock in a function.

Reply:
Fixed. Thanks